### DISCUSSION TITLE
possible to add unsupported formulas by skipping formula parsing?

### DISCUSSION THREAD

**Discussion by aliber4079** [PHPOffice/PhpSpreadsheet#3958] [Q&A]
I saw chooserows/cols isn't supported but is there a way to enter unsupported formulas without them being interpreted by PhpSpreadsheet that I could later manually calculate by opening the spreadsheet in Excel?

**Comment by MarkBaker** [comment]
The Writer will attempt to autocalculate before saving unless you disable that option using the Writer's `setPreCalculateFormulas()` method. However, if you have any auto-calculated column widths, then these require calculating all cell formulae to determine the column width, so you also need to avoid auto-sized columns.

**Reply by aliber4079** [reply]
It keeps changing =CHOOSECOLS( .. to =@CHOOSECOLS(.. even if I use setCellValueExplicit with DataType::TYPE_FORMULA

I know chooscols is unsupported but was hoping there is a way to get PhpSpreadsheet to accept the formula without evaluating or otherwise modifying it.

**Comment by MarkBaker** [comment]
PhpSpreadsheet will not change the formula that you set, it's Excel itself doing that

**Comment by aliber4079** [comment]
Even when I do something simple like this I still get the @ symbol.  I know choosecols/chooserows is unsupported, but doesn't that mean I can still use unsupported formulas, it just won't precalculate the values for me?
```<?php
require("vendor/autoload.php");
use \PhpOffice\PhpSpreadsheet\Spreadsheet;
use \PhpOffice\PhpSpreadsheet\Cell\DataType;
use \PhpOffice\PhpSpreadsheet\IOFactory;
$spreadsheet = new Spreadsheet();
$worksheet=$spreadsheet->getActiveSheet();
$worksheet->setCellValueExplicit("C1",'=CHOOSECOLS(A1:A2,1)',DataType::TYPE_FORMULA);
$writer = IOFactory::createWriter($spreadsheet, "Xlsx");
$writer->setPreCalculateFormulas(false);
$writer->save("choosecols.xlsx");
```
Extracting two sheets where one works and one doesn't, I get this (edited for clarity so may not be 100% accurate)

generated by Excel native (works):
```
<dimension ref="C1:C2"/>
<sheetViews><sheetView workbookViewId="0"><selection activeCell="C1" sqref="C1"/></sheetView></sheetViews>
<sheetData>
<row r="1" spans="3:3" x14ac:dyDescent="0.25">
<c r="C1" cm="1">
<f t="array" ref="C1:C2">_xlfn.CHOOSECOLS(A1:A2,1)</f>
<v>0</v>
</c>
</row>
</sheetData>
```

generated by PhpSpreadsheet (gets the @ sign)

```
<dimension ref="C1"/>
<sheetViews><sheetView workbookViewId="0"><selection activeCell="C1" sqref="C1"/></sheetView></sheetViews>
<sheetData>
<row r="1" spans="3:3" x14ac:dyDescent="0.25">
<c r="C1" t="e">
<f ca="1">_xludf.CHOOSECOLS(A1:A2,1)</f>
<v>#NAME?</v>
</c>
</row>
</sheetData>
```

**Comment by oleibman** [comment]
Based on the work I'm doing in PR #3962, you need to tell Excel in the XML the range of output cells to which an array formula like CHOOSECOLS will be applied in order to avoid the at sign. So, unfortunately it isn't possible to add unsupported formula support in this way. However, if I get that PR to a state where it can be implemented (possibly not far off), the good news is that it should then be relatively easy to implement CHOOSECOLS and CHOOSEROWS.

**Reply by oleibman** [reply]
Thinking more about this, it should at least be possible for Xlsx Reader to determine the dimensions of the array function results. So that, if you load a spreadsheet with a CHOOSECOLS function and save it to another spreadsheet, the dimensions will be properly preserved in the output file. I have added that ability to my PR.
